contracts/VestingEscrow.vy

```diff
@@ -0,0 +1,340 @@
+token: public(ERC20)
+start_time: public(uint256)
+end_time: public(uint256)
+cliff_length: public(uint256)
+factory: public(address)
+total_locked: public(uint256)
+
+total_claimed: public(uint256)
+disabled_at: public(uint256)
+initialized: public(bool)
+
+owner: public(address)
+manager: public(address)
+
+
+@external
+def __init__():
+    """
+    @notice Initialize source contract implementation.
+    """
+    # ensure that the original contract cannot be initialized
+    self.initialized = True
+
+
+@external
+@nonreentrant("lock")
+def initialize(
+    token: address,
+    amount: uint256,
+    recipient: address,
+    owner: address,
+    manager: address,
+    start_time: uint256,
+    end_time: uint256,
+    cliff_length: uint256,
+    factory: address,
+) -> bool:
+    """
+    @notice Initialize the contract.
+    @dev This function is separate from `__init__` because of the factory pattern
+         used in `VestingEscrowFactory.deploy_vesting_contract`. It may be called
+         once per deployment.
+    @param token Address of the ERC20 token being distributed
+    @param amount Amount of the ERC20 token to be controleed by escrow
+    @param recipient Address to vest tokens for
+    @param owner Address of the vesting owner
+    @param manager Address of the vesting manager
+    @param start_time Epoch time at which token distribution starts
+    @param end_time Time until everything should be vested
+    @param cliff_length Duration after which the first portion vests
+    @param factory Address of the parent factory
+    """
+    assert not self.initialized, "can only initialize once"
+    self.initialized = True
+
+    self.owner = owner
+    self.manager = manager
+    self.token = ERC20(token)
+    self.start_time = start_time
+    self.end_time = end_time
+    self.cliff_length = cliff_length
+
+    assert self.token.transferFrom(
+        msg.sender, self, amount
+    ), "could not fund escrow"
+
+    self.total_locked = amount
+    self.recipient = recipient
+    self.disabled_at = end_time  # Set to maximum time
+    self.factory = factory
+    log Fund(recipient, amount)
+
+    return True
+
+
+@internal
+@view
+def _total_vested_at(time: uint256 = block.timestamp) -> uint256:
+    start: uint256 = self.start_time
+    end: uint256 = self.end_time
+    locked: uint256 = self.total_locked
+    if time < start + self.cliff_length:
+        return 0
+    return min(locked * (time - start) / (end - start), locked)
+
+
+@internal
+@view
+def _unclaimed(time: uint256 = block.timestamp) -> uint256:
+    return self._total_vested_at(time) - self.total_claimed
+
+
+@external
+@view
+def unclaimed() -> uint256:
+    """
+    @notice Get the number of unclaimed, vested tokens for recipient
+    """
+    # NOTE: if `revoke_unvested` is activated, limit by the activation timestamp
+    return self._unclaimed(min(block.timestamp, self.disabled_at))
+
+
+@internal
+@view
+def _locked(time: uint256 = block.timestamp) -> uint256:
```

Comment:
Maybe we can remove this method and stay only one public method `locked()`?

Reply:
We can not call external methods inside the contract. But we need to call `_locked ` in the `revoke_all`

Reply:
You are right. Sorry, forgot that Vyper doesn't have a `public` modifier for methods.